Add section on external libraries (#713)
JuliaParallel/MPI.jl: docs/src/external.md

# External libraries and packages

Other libraries and packages may also make use of MPI. There are several concerns to ensure things are set up correctly.

## Binary requirements

> You will need to ensure that the binaries are built correctly. In particular, if you are using [using a system-provided MPI backend](@ref using_system_mpi), you will also need to use the system provided binary for the package in question.

Reviewer: ```suggestion
You need to ensure that external libraries are built correctly. In particular, if you are [using a system-provided MPI backend](@ref using_system_mpi) in Julia, you also need to use the *same* system-provided binary for all packages and external libraries you use.
```
Clarify the "single ABI" idea?

> ## Passing handles via `ccall`

Reviewer: ```suggestion
## Passing MPI handles via `ccall`
```
For clarification?

> When passing MPI.jl handle objects ([`MPI.Comm`](@ref), [`MPI.Info`](@ref), etc) to C/C++ functions [via `ccall`](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/), you should pass the object directly as an argument, and specify the argument type as either the underlying handle type (`MPI.API.MPI_Comm`, `MPI.API.MPI_Info`, etc.), or a pointer (`Ptr{MPI.API.MPI_Comm}`, `Ptr{MPI.API.MPI_Info}`, etc.). This will internally handle the unwrapping, but ensure that a reference is kept to avoid premature garbage collection.

Reviewer: ```suggestion
When passing MPI.jl handle objects ([`MPI.Comm`](@ref), [`MPI.Info`](@ref), etc) to C/C++ functions [via `ccall`](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/), you should pass the object directly as an argument, and specify the argument type as either the underlying handle type (`MPI.MPI_Comm`, `MPI.API.MPI_Info`, etc.), or a pointer (`Ptr{MPI.MPI_Comm}`, `Ptr{MPI.MPI_Info}`, etc.). This will internally handle the unwrapping, but ensure that a reference is kept to avoid premature garbage collection.
```
In my experience it works well without the `API` part. If we require the `API` part for sanity/future compatibility, I think we should state it somewhere prominently which parts of MPI.jl are expected to be called via `API` and which are just `MPI.xxx`. If there's no hard requirement yet, I'd prefer to not use the `API` part to avoid confusion. If it's just that some commonly used types are used via `MPI.xxx` and other, less commonly used or low-level entities are available only via `MPI.API.yyy` it should be documented as well.

Author: @vchuravy @giordano what do you think?

Reviewer 2: Probably I agree it's nice to keep `MPI_Comm`  part of the "high-level" API (i.e. within `MPI`, not `MPI.API`).

For example the C function signatures
```C
int cfunc1(MPI_Comm comm);
int cfunc2(MPI_Comm * comm);
```
would be called as
```julia

> ccall((:cfunc1, lib), Cint, (MPI.API.MPI_Comm,), comm)
> ccall((:cfunc2, lib), Cint, (Ptr{MPI.API.MPI_Comm},), comm)

Reviewer: ```suggestion
ccall((:cfunc1, lib), Cint, (MPI.MPI_Comm,), comm)
ccall((:cfunc2, lib), Cint, (Ptr{MPI.MPI_Comm},), comm)
```

```

## Object finalizers and `MPI.Finalize`

> External libraries may maintain allocate their own MPI handles (e.g.  create or duplicate MPI communictors), which need to be cleaned up. If these are attached to [object finalizers](https://docs.julialang.org/en/v1/base/base/#Base.finalizer), they may not be guaranteed to be called before `MPI.Finalize`, which can result in an error upon program exit. (By default, MPI.jl will install an [`atexit`](https://docs.julialang.org/en/v1/base/base/#Base.atexit) hook that calls `MPI.Finalize` if it hasn't already been invoked.)

Reviewer: ```suggestion
External libraries may allocate their own MPI handles (e.g., create or duplicate MPI communictors), which need to be cleaned up before MPI is finalized. If these are attached to [object finalizers](https://docs.julialang.org/en/v1/base/base/#Base.finalizer), they may not be guaranteed to be called before `MPI.Finalize`, which can result in an error upon program exit. (By default, MPI.jl will install an [`atexit`](https://docs.julialang.org/en/v1/base/base/#Base.atexit) hook that calls `MPI.Finalize` if it hasn't already been invoked.)
```
Sentence structure; some suggestion for clarification

There are two typical solutions to this problem:

> 1. Gate the clean up functions behind an [`MPI.Finalized](@ref) call, e.g.

Reviewer 2: ```suggestion
1. Gate the clean up functions behind an [`MPI.Finalized`](@ref) call, e.g.
```

For example the C function signatures
```C
int cfunc1(MPI_Comm comm);
int cfunc2(MPI_Comm * comm);
```
would be called as
```julia
ccall((:cfunc1, lib), Cint, (MPI.MPI_Comm,), comm)
ccall((:cfunc2, lib), Cint, (Ptr{MPI.MPI_Comm},), comm)
```

## Object finalizers and `MPI.Finalize`

External libraries may allocate their own MPI handles (e.g., create or duplicate MPI communictors), which need to be cleaned up before MPI is finalized. If these are attached to [object finalizers](https://docs.julialang.org/en/v1/base/base/#Base.finalizer), they may not be guaranteed to be called before `MPI.Finalize`, which can result in an error upon program exit. (By default, MPI.jl will install an [`atexit`](https://docs.julialang.org/en/v1/base/base/#Base.atexit) hook that calls `MPI.Finalize` if it hasn't already been invoked.)

There are two typical solutions to this problem:

1. Gate the clean up functions behind an [`MPI.Finalized`](@ref) call, e.g.

   ```julia
   finalizer(obj) do obj
       if !MPI.Finalized

>            # call clean up funciton

Reviewer 2: ```suggestion
           # call clean up function
```

       end
   end
   ```

2. Keep track of all such objects, clean them up via [`MPI.add_finalize_hook!`](@ref), e.g.

   ```julia
   finalizer(obj) do obj

>        # call clean up funciton

Reviewer 2: ```suggestion
       # call clean up function
```

   end
   MPI.add_finalize_hook!(() -> finalize(obj))
   ```
   A variant of this is to keep track of all such objects, for example, using a [`WeakKeyDict`](https://docs.julialang.org/en/v1/base/collections/#Base.WeakKeyDict), and use a hook to clean them all:
   ```julia
   const REFS = WeakKeyDict{ObjType, Nothing}()
   MPI.add_finalize_hook!() do
       for obj in keys(REFS)
           finalize(obj)
       end
   end

   # for each object `obj`
   finalizer(obj) do obj

>        # call clean up funciton

Reviewer 2: ```suggestion
       # call clean up function
```
